src/electos/ballotmaker/data/extractor.py

```diff
@@ -0,0 +1,311 @@
+        """Get the name of a candidate as it appears on a ballot."""
+        name = _text_content(candidate.ballot_name)
+        return name
+
+
+    def _candidate_party(self, candidate: Candidate):
+        """Get the name and abbreviation of the party of a candidate as it appears on a ballot.
+
+        Drop either field from result if it isn't present.
+        """
+        # Note: party ID is returned to allow de-duplicating parties in callers.
+        id_ = candidate.party_id
+        party = self._index.by_id(id_)
+        name = _text_content(party.name) if party else None
+        abbreviation = (
+            _text_content(party.abbreviation) if party and party.abbreviation else None
+        )
+        result = {}
+        if name:
+            result["name"] = name
+        if abbreviation:
+            result["abbreviation"] = abbreviation
+        return result, id_
+
+
+    def _candidate_contest_candidates(self, contest: CandidateContest):
+        """Get candidates for contest, grouped by slate/ticket.
+
+        A slate has:
+
+        - A single ID for the contest selection
+        - Collects candidate names into an array.
+        - Collects candidate parties into an array.
+          - If all candidates in a race share a single party they are combined into
+            one entry in the array.
+          - If any candidates differ from the others, parties are listed separately.
+
+        Notes:
+            - There's no clear guarantee of a 1:1 relationship between slates/tickets
+              and parties.
+        """
+        # Collect individual candidates
+        candidates = []
+        for selection in contest.contest_selection:
+            assert isinstance(selection, CandidateSelection), \
+                f"Unexpected non-candidate selection: {type(selection).__name__}"
+            names = []
+            parties = []
+            _party_ids = set()
+            if selection.candidate_ids:
+                for id_ in selection.candidate_ids:
+                    candidate = self._index.by_id(id_)
+                    name = self._candidate_name(candidate)
+                    if name:
+                        names.append(name)
+                    party, _party_id = self._candidate_party(candidate)
+                    if party:
+                        parties.append(party)
+                        _party_ids.add(_party_id)
+            # If there's only one party ID, all candidates share the same party.
+            # If there's any divergence track them all individually.
+            if len(_party_ids) == 1:
+                parties = parties[:1]
+            result = {
+                "id": selection.model__id,
+                "name": names,
+                "party": parties,
+                "is_write_in": bool(selection.is_write_in)
+            }
+            candidates.append(result)
+        return candidates
+
+
+    def _candidate_contest_offices(self, contest: CandidateContest):
```

Comment:
Hmm. That could work. I'm leaning towards keeping them different primarily because we're changing the shape of this data as we go, and if we decide that parties behave differently than offices we have to split them up again. But I'll think about it.